src/main/kotlin/biokotlin/cli/MutateProteins.kt

```diff
@@ -0,0 +1,363 @@
+
+        ranges
+            .sortedBy { it.start }
+            .forEach { range ->
+                if (result.isEmpty()) {
+                    result.add(range)
+                } else {
+                    val last = result.last()
+                    if (last.end >= range.start) {
+                        result[result.size - 1] = BedfileRecord(last.contig, last.start, range.end)
+                    } else {
+                        result.add(range)
+                    }
+                }
+            }
+
+        return result
+
+    }
+
+    // determine if index is in any of the ranges
+    private fun inRanges(index: Int, ranges: Collection<BedfileRecord>): Boolean {
+        return ranges.any { it.contains(index) }
+    }
+
+    // All ranges should have the same contig
+    // Ranges should not overlap
+    private fun inverseRanges(
+        contig: String,
+        seqLength: Int,
+        ranges: Collection<BedfileRecord>
+    ): Collection<BedfileRecord> {
+
+        if (ranges.isEmpty()) {
+            return listOf(BedfileRecord(contig, 0, seqLength - 1))
+        }
+
+        val result = mutableListOf<BedfileRecord>()
+
+        var start = 0
+        ranges
+            .sortedBy { it.start }
+            .forEach { range ->
+                if (range.start > start) {
+                    result.add(BedfileRecord(contig, start, range.start - 1))
+                }
+                start = range.end + 1
+            }
+        if (start < seqLength) {
+            result.add(BedfileRecord(contig, start, seqLength - 1))
+        }
+
+        return result
+
+    }
+
+    /**
+     * This deletes bases (number specified by parameter length) from each sequence.
+     * The bases are continuous. It makes the deletion in or out of
+     * the ranges defined by the bedfile depending on the setting of putMutationsInRanges.
+     */
+    private fun deletionMutation(
+        record: ProteinSeqRecord, ranges: Multimap<String, BedfileRecord>
+    ): String {
+
+        val origSeq = record.seq()
+        val seqLength = origSeq.length
+
+        val contig = record.id
+        val rangesForRecord = ranges.get(contig)
+
+        val mergedRanges = validateAndMergeRanges(seqLength, rangesForRecord)
+
+        val possibleMutationRanges =
+            if (putMutationsInRanges) mergedRanges else inverseRanges(contig, seqLength, mergedRanges)
+
+        // get possible ranges to mutate
+        // that are long enough to delete specified
+        // length. Then randomly select one of these ranges
+        val rangeToMutate = possibleMutationRanges
+            .filter { range -> range.length() >= length }
+            .randomOrNull()
+
+        val result = StringBuilder(origSeq)
+
+        if (rangeToMutate != null) {
+            // get random start index of place to delete
+            // within rangeToMutate
+            val start = Random(4321).nextInt(rangeToMutate.start, rangeToMutate.end - length + 2)
```

Comment:
Besides removing hard-coded seed numbers, perhaps explain what `+ 2` is doing here in a comment?

Reply:
Ok, I added a comment about the +2